### ISSUE TITLE
S2 compression for Parquet files

### ISSUE THREAD

**Issue by marklit** [ClickHouse/ClickHouse#45130]
ClickHouse is able to produce Parquet files with Snappy compression atm but not with S2. S2, the extension to Snappy, can compress at 16x the rate of regular Snappy while achieving a higher ratio.

The above would make transporting data between CilckHouse instances at the least a lot faster than regular Parquet. At some point, there will be more software supporting S2 down the road as Parquet has had many codecs added to it over the years.

Klaus Post wrote an implementation in [GoLang and Assembler](https://github.com/klauspost/compress/blob/master/s2/_generate/gen.go) which is compiled into a [.so file](https://github.com/klauspost/compress/blob/master/s2/encodeblock_amd64.s) with https://github.com/mmcloughlin/avo. The Assembler itself is very vanilla Intel and ARM, [no AVX2](https://github.com/klauspost/compress/discussions/732), etc...

Since ClickHouse would be the first Parquet writer that supports S2 that I know of you'd probably get to decide what the magic identifier is.

**Comment by alexey-milovidov** [comment]
It's Go assembly syntax, not AT&T (gas) or Intel (nasm), and it is using Go calling conventions, for example:
```
TEXT ·encodeBlockAsm(SB), $65560-56
        MOVQ dst_base+0(FP), AX
        MOVQ $0x00000200, CX
        LEAQ 24(SP), DX
        PXOR X0, X0
```

Note this `·` symbol.

Nevertheless, it is possible to convert it to another syntax if needed by compiling and producing an `objdump` then.

> can compress at 16x the rate of regular Snappy

In fact, it is only up to 1.9 times faster, typically just 1.1 times faster.
Everything else applies the algorithm in parallel, which is trivial for Snappy as well.

Snappy is not the best compression format compared to LZ4.
Switching to LZ4 is expected to give better performance benefits.

A more concerning issue is: the library is not continuously fuzzed: https://github.com/klauspost/compress/issues/289
(this is a must for every compression library)

The last time a memory safety bug was found for S2 was just 1.5 years ago: https://github.com/klauspost/compress/pull/362/files - it makes potential usage of the library risky.

TLDR: better to simply use LZ4 in Parquet instead of Snappy.

**Comment by alexey-milovidov** [comment]
Will be addressed by this PR: #47114
